File: lc39.java
class Solution {
    List<List<Integer> > ans=new ArrayList<>();

    public void dfs(int i,int sum,List<Integer> currlist,int[] candidates){
        if(i==candidates.length && sum==0){
            ans.add(new ArrayList<>(currlist));
            
        }

        if(i>=candidates.length || sum<0){
            return;
        }

        currlist.add(candidates[i]);
        dfs(i,sum-candidates[i],currlist,candidates);

        currlist.remove(currlist.size()-1);

        dfs(i+1,sum,currlist,candidates);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        List<Integer> currlist=new ArrayList<>();
        dfs(0,target,currlist,candidates);
        return ans;
    }
}

File: lc39.cpp
class Solution {
    vector<vector<int>> ans;

public:

    void dfs(int i,int sum,vector<int>& candidates, vector<int>& currlist){
        if(i==candidates.size() && sum==0){
            ans.push_back(currlist);
        }

        if(sum<0 || i>=candidates.size()){
            return;
        }

        currlist.push_back(candidates[i]);

        dfs(i,sum-candidates[i],candidates,currlist);

        currlist.pop_back();

        dfs(i+1,sum,candidates,currlist);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {

      vector<int> currlist;
       dfs(0,target,candidates,currlist);
        return ans;
        
    }
};
